/ACM-3910 Swope pull model revised (#5158)
open-cluster-management/rhacm-docs: applications/gitops_push_pull.adoc

[#gitops-push-pull]
= Deploying Argo CD with the Push and Pull model

Using a _Push model_, The Argo CD server on the hub cluster deploys the application resources on the managed clusters. For the _Pull model_, the application resources are propagated by the _Propagation controller_ to the managed clusters by using `manifestWork`. 

> For both models, the same `ApplicationSet` CRD is used to deploy the application to the managed cluster.

Reviewer: ```suggestion
For both models, the same `ApplicationSet` custom resource definition is used to deploy the application to the managed cluster.
```

Author: we cannot say CRD?

*Required access:* Cluster administrator

> *Information:* If your `openshift-gitops-ArgoCD-application-controller` service account is _not_ assigned as a cluster administrator, the GitOps application controller might not deploy resources. The application status might send an error similar to the following error:

Reviewer: ```suggestion
*Note:* If your `openshift-gitops-ArgoCD-application-controller` service account is _not_ assigned as a cluster administrator, the GitOps application controller might not deploy resources. The application status might send an error similar to the following error:
```

Reviewer: Information is not on our current list that we use: https://stolostron.github.io/rhacm-docs/wr_special_tag_guidance.

Author: I think I meant to use Important

----
cannot create resource "services" in API group "" in the namespace
"mortgage",deployments.apps is forbidden: User
"system:serviceaccount:openshift-gitops:openshift-gitops-Argo CD-application-controller"
----

If you are not a cluster administrator and need to resolve this issue, complete the following steps:

. Create all namespaces on each managed cluster where the Argo CD application will be deployed.

. Add the `managed-by` label to each namespace. If an Argo CD application is deployed to multiple namespaces,
each namespace should be managed by Argo CD.

+
See the following example with the `managed-by` label:

+
[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: mortgage2
  labels:
    argocd.argoproj.io/managed-by: openshift-gitops
----

. You must declare all application destination namespaces in the repository for the application and include the `managed-by` label in the namespaces. Refer to _Additional resources_ to learn how to declare a namespace.
//still need a good link 6/8

[#push-pull-arch]
== Argo CD Push and Pull model architecture

> For both Push and Pull model, the _Argo CD ApplicationSet controller_ on the hub cluster reconciles to create application resources for each target managed cluster. See more information about architecture for both models:

Reviewer: ```suggestion
For both Push and Pull model, the _Argo CD ApplicationSet controller_ on the hub cluster reconciles to create application resources for each target managed cluster. See the following information about architecture for both models:
```

- For Pull model, an Argo CD server must be running on each target managed cluster. The Argo CD application resources are replicated on the managed clusters, which are then deployed by the local Argo CD server. The distributed Argo CD applications on the managed clusters are created with a single Argo CD `ApplicationSet` resource on the hub cluster.

- The managed cluster is determined by the value of the `ocm-managed-cluster` annotation.

- For successful implementation of Pull model, the Argo CD application controller must ignore Push model application resources with the `argocd.argoproj.io/skip-reconcile` annotation in the template section of the `ApplicationSet`.

- For Pull model, the _Argo CD Application controller_ on the managed cluster reconciles to deploy the application.

- The Pull model _Resource sync controller_ on the hub cluster queries the {ocm} search V2 component on each managed cluster periodically to retrieve the resource list and error messages for each Argo CD application.

> - The _Aggregation controller_ on the hub cluster creates and updates the `MulticlusterApplicationSetReport` from across clusters by using the data from the Resource sync controller, and the status information from the `manifestWork`.

Reviewer: ```suggestion
- The _Aggregation controller_ on the hub cluster creates and updates the `MulticlusterApplicationSetReport` from across clusters by using the data from the Resource sync controller, and the status information from `manifestWork`.
```

[#prereqs-pull-model]
== Prerequisites 

See the following prerequisites to use the Argo CD pull model:

> - The GitOps operator must be installed on the hub cluster and the target managed clusters in the `openshift-gitops` namespace. *Note* Use the {ocp-short} OperatorHub to install the GitOps operator, where the default namespace is `openshift-gitops`.

Reviewer: ```suggestion
- The GitOps operator must be installed on the hub cluster and the target managed clusters in the `openshift-gitops` namespace. *Note:* Use the {ocp-short} OperatorHub to install the GitOps operator, where the default namespace is `openshift-gitops`.
```

//developer to get back to us here... Installing this way on managed clusters is a lot if there are many; consider limitation if we keep it like this or add a policy. bcs 6/5

- The required hub cluster {ocp-short} GitOps operator must be version 1.9.0 or later. 

- The required managed clusters {ocp-short} GitOps operator must be the same version as the hub cluster.

- You need the _ApplicationSet controller_ to propagate the Argo CD application template for a managed cluster.

> - Every managed cluster must have a cluster secret in the Argo CD server namespace on the hub cluster, which is required by the ArgoCD application set controller to propagate the Argo CD application template for a managed cluster. Create a `gitOpsCluster` resource that contains a reference to a `placement` resource. The `placement` resource selects all the managed clusters that need to support the pull model so that managed cluster secrets are created in the Argo CD server namespace.

Reviewer: ```suggestion
- Every managed cluster must have a cluster secret in the Argo CD server namespace on the hub cluster, which is required by the ArgoCD application set controller to propagate the Argo CD application template for a managed cluster. Create a `gitOpsCluster` resource that contains a reference to a `placement` resource. The `placement` resource selects all of the managed clusters that need to support the pull model, so that managed cluster secrets are created in the Argo CD server namespace.
```

> - The GitOps cluster resource must contains a reference to a `placement` resource, since the `placement` resource selects all the managed clusters that support the pull model. With this prerequisite, the managed cluster secrets are created in the Argo CD server namespace.

Reviewer: ```suggestion
- The GitOps cluster resource must contains a reference to a `placement` resource, since the `placement` resource selects all of the managed clusters that support the pull model. With this prerequisite, the managed cluster secrets are created in the Argo CD server namespace.
```

[#crd-pull-model]
== Creating the _ApplicationSet_ CRD

> The Argo CD `ApplicationSet` CRD is used to deploy applications on the managed clusters by using the Push or Pull model with a `placement` resource in the generator field that is used to get a list of managed clusters. 

Reviewer: I think in general, we are spelling out "custom resource definition". In the title, it might be better to use the acronym for space.

. For the Pull model, set the destination for the application to the default local Kubernetes server, as displayed in the following example. The application is deployed locally by the application controller on the managed cluster. 

. Add the annotations that are required to override the default Push model, as displayed in the following example `ApplicationSet` YAML, which uses the pull model with template annotations:

+
[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: `ApplicationSet`
metadata:
  name: guestbook-allclusters-app-set
  namespace: openshift-gitops
spec:
  generators:
  - clusterDecisionResource:
      configMapRef: ocm-placement-generator
      labelSelector:
        matchLabels:
          cluster.open-cluster-management.io/placement: aws-app-placement
      requeueAfterSeconds: 30
  template:
    metadata:
      annotations:
        apps.open-cluster-management.io/ocm-managed-cluster: '{{name}}'<1>
        apps.open-cluster-management.io/ocm-managed-cluster-app-namespace: openshift-gitops 

>         argocd.argoproj.io/skip-reconcile: "true" <3>

Reviewer: Is there a reason that 3 is between 2 and 1?

Author: No, last minute edit from dev and I forgot to rearrange. good catch.

  - The overall sync and health status for each Argo CD application
  - An error message for each cluster where the overall status is `out of sync` or `unhealthy`
  - A summary status all the states of your managed clusters

- The _Resource sync controller_ and the _Aggregation controller_ both run every 10 seconds to create the report.

- The two controllers, along with the Propagation controller, run in separate containers in the same `multicluster-integrations` pod, as shown in the following example output:

+
----
NAMESPACE               NAME                                       READY   STATUS  
open-cluster-management multicluster-integrations-7c46498d9-fqbq4  3/3     Running  
----

The following is an example `MulticlusterApplicationSetReport` YAML file for the `guestbook` application:

[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1alpha1
kind: MulticlusterApplicationSetReport
metadata:
  labels:
    apps.open-cluster-management.io/hosting-applicationset: openshift-gitops.guestbook-allclusters-app-set
  name: guestbook-allclusters-app-set
  namespace: openshift-gitops
statuses:
  clusterConditions:
  - cluster: cluster1
    conditions:

>     - message: 'Failed sync attempt to 53e28ff20cc530b9ada2173fbbd64d48338583ba: one or more objects failed to apply, reason: services is forbidden: User "system:serviceaccount:openshift-gitops:openshift-gitops-Argo CD-application-controller" cannot create resource "services" in API group "" in the namespace "guestbook",deployments.apps is forbidden: User "system:serviceaccount:openshift-gitops:openshift-gitops-Argo CD-application-controller" cannot create resource "deployments" in API group "apps" in the namespace "guestboo...'

Author: @philipwu08 I feel we need to look at line 182 and give a more generic example of a failed message?

      type: SyncError
    healthStatus: Missing
    syncStatus: OutOfSync
  - cluster: pcluster1
    healthStatus: Progressing
    syncStatus: Synced
  - cluster: pcluster2
    healthStatus: Progressing
    syncStatus: Synced
  summary:
    clusters: "3"
    healthy: "0"
    inProgress: "2"
    notHealthy: "3"
    notSynced: "1"
    synced: "2"
----

*Note:* If a resource fails to deploy, the resource is not included in the resource list. See error messages for information.

[#pull-push-resources]
== Additional resources

 - https://docs.openshift.com/container-platform/4.11/cicd/gitops/configuring-an-openshift-cluster-by-deploying-an-application-with-cluster-configurations.html#creating-an-application-by-using-the-oc-tool_configuring-an-openshift-cluster-by-deploying-an-application-with-cluster-configurations 

> //working for either a replacement (not easily in 4.13) 6/5

Author: still working on good links here

Reviewer: Maybe this: https://access.redhat.com/documentation/en-us/openshift_container_platform/4.13/html/cicd/gitops#configuring-an-openshift-cluster-by-deploying-an-application-with-cluster-configurations

Author: I gave it to them in the Google doc, we just have to connect of they want it.
